mock_client.go

```diff
@@ -16,7 +16,9 @@ import (
 type MockClient struct {
 	root string
 
-	Err error
+	Err                 error
+	Calls               int // count number of calls to Open(), Delete(), UploadFile(), ListFiles(), Walk()
+	DesiredErrResponses int // stop returning Err after this many calls to Open(), Delete(), UploadFile(), ListFiles(), Walk()
```

Comment:
Can we make these names a bit more clear? It's not quite clear what those names refer to and they seem to tie us to initial failures, rather than spurious failures. 

Are we trying to inject random failures in client calls? I've done similar elsewhere with a bit different pattern.

https://github.com/moov-io/achgateway/blob/master/internal/incoming/stream/streamtest/streamtest.go#L56
https://github.com/moov-io/achgateway/blob/master/internal/pipeline/file_receiver.go#L83

https://github.com/moovfinancial/rtpgateway/blob/master/internal/rtp/mock_message_handler.go

Reply:
Im not trying to simulate random or spurious failures. Im trying to test retry logic that im writing in mc-settlement. In order to test it, I need to have mock_client be able to return errors for the first X number of calls, and then no error on subsequent calls. Where X = DesiredErrResponses.
My test will assert how many times the mock_client was called/retried

Reply:
or it might make more sense for mock_client to define an interface, and my code can implement a mock using that interface. Ill play around with it tomorrow..

Reply:
That makes sense, but it lets your code off the hook from being tested for errors after the initial few. Your code is probably iterating over, so initial/ongoing failures likely represent themselves the same. 

Can we change those names to something like `OperationsCalled` and `RemainingErrors`? I don't quite like how vague the current names are.

Reply:
@adamdecaf Im not sure what you mean by that first part of your last comment.

Reply:
Assuming calls fail in a pattern like: error, error, error, success, success is only one way they could occur. I'm assuming your code is getting a list of files and iterating over them. One of those iterations could fail and the code should be hardened against that. Only having retry logic to handle "error, error, error, success" in the List() call would leave the `for` loop untested.

Reply:
In achgateway we randomly inject failures into the mix, which helps to harden the iterative loop. 

https://github.com/moov-io/achgateway/blob/master/internal/test/upload_test.go#L197-L202 

I'm not saying your code has to randomly inject failures, but this addition only handles one type of error. It's very common to have 1 of 100 .Open calls fail and your code should be tolerant of such.

Reply:
my changes in this PR are only intended to simulate a series of error responses followed by a success response. This is needed to test my retry logic. Im **not** trying to test something like error, success, error, success, or any random combo of error/success responses.

Also, I could move retry logic from my microservice into go-sftp